sources/core/Stride.Core.Mathematics/Matrix.cs

```diff
@@ -1014,39 +1017,42 @@ public static void Multiply(ref Matrix left, ref Matrix right, out Matrix result
-            result.M12 = (left.M11 * right.M12) + (left.M12 * right.M22) + (left.M13 * right.M32) + (left.M14 * right.M42);
-            result.M22 = (left.M21 * right.M12) + (left.M22 * right.M22) + (left.M23 * right.M32) + (left.M24 * right.M42);
-            result.M32 = (left.M31 * right.M12) + (left.M32 * right.M22) + (left.M33 * right.M32) + (left.M34 * right.M42);
-            result.M42 = (left.M41 * right.M12) + (left.M42 * right.M22) + (left.M43 * right.M32) + (left.M44 * right.M42);
-            result.M13 = (left.M11 * right.M13) + (left.M12 * right.M23) + (left.M13 * right.M33) + (left.M14 * right.M43);
-            result.M23 = (left.M21 * right.M13) + (left.M22 * right.M23) + (left.M23 * right.M33) + (left.M24 * right.M43);
-            result.M33 = (left.M31 * right.M13) + (left.M32 * right.M23) + (left.M33 * right.M33) + (left.M34 * right.M43);
-            result.M43 = (left.M41 * right.M13) + (left.M42 * right.M23) + (left.M43 * right.M33) + (left.M44 * right.M43);
-            result.M14 = (left.M11 * right.M14) + (left.M12 * right.M24) + (left.M13 * right.M34) + (left.M14 * right.M44);
-            result.M24 = (left.M21 * right.M14) + (left.M22 * right.M24) + (left.M23 * right.M34) + (left.M24 * right.M44);
-            result.M34 = (left.M31 * right.M14) + (left.M32 * right.M24) + (left.M33 * right.M34) + (left.M34 * right.M44);
-            result.M44 = (left.M41 * right.M14) + (left.M42 * right.M24) + (left.M43 * right.M34) + (left.M44 * right.M44);
+            unsafe
+            {
+                fixed (Matrix* outPtr = &result)
+                {
+                    ref MatrixDotnet l = ref UnsafeRefAsDotNet(in left);
+                    ref MatrixDotnet r = ref UnsafeRefAsDotNet(in right);
+                    *(MatrixDotnet*)outPtr = LayoutIsRowMajor ? l * r : r * l;
+                    return; // Explicit return until JIT optimization is fixed
+                }
+            }
         }
 
         /// <summary>
         /// Determines the product of two matrices.
+        /// Variables passed as <paramref name="left"/> or <paramref name="right"/> must not be used as the out parameter
+        /// <paramref name="result"/>, because <paramref name="result"/> is calculated in-place.
         /// </summary>
         /// <param name="left">The first matrix to multiply.</param>
         /// <param name="right">The second matrix to multiply.</param>
-        /// <returns>The product of the two matrices.</returns>
-        public static Matrix Multiply(Matrix left, Matrix right)
+        /// <param name="result">The product of the two matrices.</param>
+        [Obsolete($"Use {nameof(Multiply)} instead")]
+        public static void MultiplyRef(ref Matrix left, ref Matrix right, ref Matrix result)
         {
-            Matrix result;
             Multiply(ref left, ref right, out result);
-            return result;
         }
```

Comment:
Small diff fail here, `Matrix Multiply(Matrix left, Matrix right)` hasn't been removed and `MultiplyRef` isn't new, the former is now right below and the later was above but has been replaced.